DOC-2386: add Google Gemini integration guide (#3250)
tinymce/tinymce-docs: modules/ROOT/pages/ai-gemini.adoc

:keywords: example, demo, custom, plugin, ai, assistant, guide, google, gemini

include::partial$misc/admon-ai-pricing.adoc[]

== Introduction

This guide provides instructions for integrating the {pluginname} plugin using Google Gemini in {productname}.

There are two version of the Google Gemini API: Google AI Studio and Google Cloud - Vertex AI. The Google AI Studio API is ideal for quickly integrating the Google Gemini models into an application with the minimal setup required. The Google Cloud - Vertex AI API is more powerful and flexible than Google AI Studio, but requires more setup to integrate the Google Gemini models into an application.

Vertex AI provides a `systemInstruction` field in the request body that allows controlling the output of the AI model. Google AI Studio does not currently provide this field, so the system messages are provided as part of the user prompt.

TIP: Google AI Studio and Vertex AI can both be used to generate string and streaming responses.

To learn more about the difference between string and streaming responses, see xref:ai.adoc#the-respondwith-object[The `respondWith` object] on the plugin page.

> 

Reviewer: We should include a note on each page which recommends setting up a proxy service for authentication and hiding keys, and display it before any examples are shown.

Author: @MitchC1999 We already have a comment in the code examples above `const api_key` that states this. Do you want to move that to an admonition outside the code?

Reviewer: Yea a note outside the code block would be more visible

Author: Added admonition: [93c193f](https://github.com/tinymce/tinymce-docs/pull/3250/commits/93c193f5079db62b3f83b7354ac6b6041e620bc4)

If you are happy with that, I will also add it to OpenAI and Bedrock pages.

== String response

This example demonstrates how to integrate the {pluginname} plugin with the https://ai.google.dev/[Google AI Studio] REST API to generate string responses. Google AI Studio is ideal for quickly integrating the Google Gemini models into an application with the minimal setup required.

See https://ai.google.dev/gemini-api/docs/get-started/rest[Google AI Studio Gemini API] for more information on the Google AI Studio Gemini API.

[source,js]
----
// This example stores the API key in the client side integration. This is not recommended for any purpose.
// Instead, an alternate method for retrieving the API key should be used.
const API_KEY = '<INSERT_API_KEY_HERE>';
const API_URL = `https://generativelanguage.googleapis.com/v1beta/models/gemini-pro:generateContent?key=${API_KEY}`;

const ai_request = (request, respondWith) => {
  const geminiOptions = {
	method: 'POST',
	headers: {
	  'Content-Type': 'application/json',
	},
	body: JSON.stringify({
	  contents: [{
		parts: [

> 		{ text: request.prompt }

Reviewer: ```suggestion
		    { text: request.prompt }
```
Ensure that indentation is consistent.

		]
	  }],
	  generationConfig: {
		temperature: 0.9,
		maxOutputTokens: 800,
	  }
	})
  };
  respondWith.string((signal) => window.fetch(API_URL, { signal, ...geminiOptions })
	.then(async (response) => {
	  if (response) {
		const data = await response.json();
		if (data.error) {
		  throw new Error(`${data.error.type}: ${data.error.message}`);
		} else if (response.ok) {
		  // Extract the response content from the data returned by the API

> 		  return data?.candidates[0]?.content?.parts[0]?.text?.trim().replace(/^```html\n/, "").replace(/\n```$/, "");

Reviewer: Don't need the markdown workaround here

	const contents = [
	  ...conversation,
	  { role: 'user', parts: [
		{ text }
	  ]}
	];

	const generationConfig = {
	  temperature: 0.9,
	  maxOutputTokens: 800,
	};

	const requestBody = {
	  contents,
	  generationConfig,
	  systemInstruction,
	};

	const geminiOptions = {
	  method: 'POST',
	  headers: {
		'Content-Type': 'application/json',
		'Authorization': `Bearer ${ACCESS_TOKEN}`
	  },
	  body: JSON.stringify(requestBody)
	};

	const onopen = async (response) => {
	  if (response) {
		const contentType = response.headers.get('content-type');
		if (response.ok && contentType?.includes('text/event-stream')) {
		  return;
		} else if (contentType?.includes('application/json')) {
		  const data = await response.json();
		  if (data.error) {
			throw new Error(`${data.error.type}: ${data.error.message}`);
		  }
		}
	  } else {
		throw new Error('Failed to communicate with the Gemini API');
	  }
	};

	// This function passes each new message into the plugin via the `streamMessage` callback.
	const onmessage = (ev) => {
	  const data = ev.data;
	  if (data !== '[DONE]') {
		const parsedData = JSON.parse(data);

> 		const message = parsedData?.candidates[0]?.content?.parts[0]?.text?.replace(/^```html\n/, "").replace(/\n```$/, "");

Reviewer: We don't need to include the markdown replacement code.

Author: I found that without the replace statements, the response contains the markdown formatting:
<img width="1222" alt="image" src="https://github.com/tinymce/tinymce-docs/assets/40410898/03ce4def-d45e-4663-9bff-e1d12e512777">

Are you okay with this?

Reviewer: That's a string response, I don't think that workaround will work in a streaming response

Author: Based on my testing, it seems to work with streaming too (see console).

<img width="1649" alt="image" src="https://github.com/tinymce/tinymce-docs/assets/40410898/1d9c50be-e6ae-438a-836e-c79a9abfdacd">

Reviewer: Looks good then. It seems Gemini consistently chunks that text in a single message, where OpenAI would often separate it into a few messages.

Author: Yes, OpenAI's responses are much more frequent and fragmented. Gemini returns much larger chunks, so the markdown formatting almost always ends up in the first and last chunks.

The access token is used in the Authorization header to authenticate the request to the Google Cloud API. The access token should be stored securely and not exposed to the client side integration. A recommended approach is to use a proxy server to generate the access token and make the API calls. For more information, see the xref:ai-proxy.adoc[AI Proxy Server reference guide].

The https://www.npmjs.com/package/google-auth-library[google-auth-library] can be used to programmatically generate an access token for the Google Cloud API. Alternatively, the https://cloud.google.com/nodejs/docs/reference/vertexai/latest[Vertex AI Node.js SDK] or other client library can be used, but may require modifications to the code.

== Prerequisites

Before you begin, you need to have the following:

* A Google Cloud Platform (GCP) account.
* A Google Cloud Project with the Vertex AI API enabled.
* The Project ID and Region of the Google Cloud Project.
* An Access Token for the Vertex AI API.

== String response

This example demonstrates how to integrate the {pluginname} plugin with the Vertex AI API to generate string responses.

include::partial$misc/admon-ai-proxy.adoc[]

[source,js]
----

> // Not recommended for production use. Use a proxy server to generate the access token.

Reviewer: ```suggestion
// Providing access credentials within the integration is not recommended for production use.
// We recommend setting up a proxy server to authenticate requests and provide access.
```

		parts: { text: request.prompt }
	  },
	  generationConfig: {
		temperature: 0.9,
		maxOutputTokens: 800,
	  }
	})
  };
  respondWith.string((signal) => window.fetch(API_URL, { signal, ...geminiOptions })
	.then(async (response) => {
	  if (response) {
		const data = await response.json();
		if (data.error) {
		  throw new Error(`${data.error.type}: ${data.error.message}`);
		} else if (response.ok) {
		  // Extract the response content from the data returned by the API
		  return data?.candidates[0]?.content?.parts[0]?.text?.trim().replace(/^```html\n/, "").replace(/\n```$/, "");
		}
	  } else {
		throw new Error('Failed to communicate with the Gemini API');
	  }
	})
  );
};

tinymce.init({
  selector: 'textarea',  // change this value according to your HTML
  plugins: 'ai',
  toolbar: 'aidialog aishortcuts',
  ai_request
});
----

== Streaming response

This example demonstrates how to integrate the {pluginname} plugin with the Vertex AI API to generate streaming responses.

include::partial$misc/admon-ai-proxy.adoc[]

[source,js]
----
const fetchApi = import("https://unpkg.com/@microsoft/fetch-event-source@2.0.1/lib/esm/index.js").then(module => module.fetchEventSource);

> // Not recommended for production use. Use a proxy server to generate the access token.

Author: ```suggestion
// Providing access credentials within the integration is not recommended for production use.
	// We recommend setting up a proxy server to authenticate requests and provide access.
```

To learn more about the difference between string and streaming responses, see xref:ai.adoc#the-respondwith-object[The `respondWith` object] on the plugin page.

== Authentication

The access token is used in the Authorization header to authenticate the request to the Google Cloud API. The access token should be stored securely and not exposed to the client side integration. A recommended approach is to use a proxy server to generate the access token and make the API calls. For more information, see the xref:ai-proxy.adoc[AI Proxy Server reference guide].

The https://www.npmjs.com/package/google-auth-library[google-auth-library] can be used to programmatically generate an access token for the Google Cloud API. Alternatively, the https://cloud.google.com/nodejs/docs/reference/vertexai/latest[Vertex AI Node.js SDK] or other client library can be used, but may require modifications to the code.

== Prerequisites

Before you begin, you need to have the following:

* A Google Cloud Platform (GCP) account.
* A Google Cloud Project with the Vertex AI API enabled.
* The Project ID and Region of the Google Cloud Project.
* An Access Token for the Vertex AI API.

> 
> == String response
> 
> This example demonstrates how to integrate the {pluginname} plugin with the Vertex AI API to generate string responses.
> 
> include::partial$misc/admon-ai-proxy.adoc[]

Reviewer: ```suggestion

include::partial$misc/admon-ai-proxy.adoc[]

== String response

This example demonstrates how to integrate the {pluginname} plugin with the Vertex AI API to generate string responses.
```
We will only need to include this partial once per page, before the example headings.

    },
    body: JSON.stringify({
      contents: {
        role: "user",
        parts: { text: request.prompt }
      },
      generationConfig: {
        temperature: 0.9,
        maxOutputTokens: 800,
      }
    })
  };
  respondWith.string((signal) => window.fetch(API_URL, { signal, ...geminiOptions })
    .then(async (response) => {
      if (response) {
        const data = await response.json();
        if (data.error) {
          throw new Error(`${data.error.type}: ${data.error.message}`);
        } else if (response.ok) {
          // Extract the response content from the data returned by the API
          return data?.candidates[0]?.content?.parts[0]?.text?.trim().replace(/^```html\n/, "").replace(/\n```$/, "");
        }
      } else {
        throw new Error('Failed to communicate with the Gemini API');
      }
    })
  );
};

tinymce.init({
  selector: 'textarea',  // change this value according to your HTML
  plugins: 'ai',
  toolbar: 'aidialog aishortcuts',
  ai_request
});
----

== Streaming response

This example demonstrates how to integrate the {pluginname} plugin with the Vertex AI API to generate streaming responses.

> 
> include::partial$misc/admon-ai-proxy.adoc[]
> 

Author: ```suggestion

```
